block-layer/content-addressable-archives.md

```diff
@@ -30,51 +30,110 @@
+
+[ varint | cbor block ] [ varint | CID | dag-cbor block ] [ varint | CID | block ] [ varint | CID | block ] …
+```
+
+To maintain backward compatibility, Version 2 retains the lead-in block without CID but _only_ includes a CBOR encoded map containing a single key/value pair: `{ version: 2 }`. Therefore, a decoder may read the first 11 bytes of a CAR and match it against the bytes: `0x0aa16776657273696f6e02` (`0x0a` is a varint representing `10`, followed by 10 bytes of the CBOR encoded form of `{ version: 2 }`).
+
+Described as an IPLD Schema:
+
+```ipldsch
+type CarVersion struct {
+  version Int
+}
+```
+
+All following blocks are prefixed by a varint describing the number of bytes to read, followed by the block's CID, followed by the block's binary data.
+
+The first block in this list is the _Header_, a DAG-CBOR encoded object, containing roots, selectors and deterministic descriptor.
+
+Described as an IPLD Schema:
+
+```ipldsch
+type CarRoot struct {
+  root Link
+  selector optional Selector   # see the Selectors specification, SelectorEnvelope not necessary
+} representation tuple
+
+type CarHeader struct {
+  deterministic optional Bool (implicit "false")
+  roots [CarRoot]
+}
+```
+
+All remaining blocks are form the Data portion of the CAR, containing the roots and blocks constituting the DAGs described by their selectors.
+
+For non-deterministic CARs it is also possible that the Data portion of the CAR contain blocks not associated with the roots presented in the Header block.
+
+#### Constraints
+
+* The `deterministic` field may be absent, indicating a value of `false`. A deterministic CAR _must_ include a `deterministic` field with a value of `true`.
+* Where a `deterministic` value of `true` is present:
+  * The `roots` array _must_ contain one or more elements.
+  * Each element within the `roots` array _must_ contain a `selector`, even if that selector describes an entire DAG under the root (`{ "a": { ">": "." } }` in encoded form).
+* Where a `deterministic` field is not present or has a value of `false`:
+  * The `roots` array may contain zero or more elements.
+  * Elements in the `roots` array may omit the `selector`, indicating a possibly-incomplete DAG.
+* Any CID present in the `roots` array _must_ describe a block present in the _Data_ section of the CAR.
 
 ### Data
 
-Immediately following the Header block, **one or more** IPLD blocks are concatenated to form the _Data_ section of the CAR format. _(Caveat: see [Zero blocks](#zero-blocks) under Unresolved Issues.)_ Each block is encoded into a _Section_ by the concatenation of the following values:
+Immediately following the Header block, **one or more** IPLD blocks are concatenated to form the _Data_ section of the CAR format. It is invalid for a CAR to contain zero blocks after the Header. Each block is encoded into a _Section_ by the concatenation of the following values:
```

Comment:
For this use case, I think we should have a header that doesn’t specify any roots, and we should figure out who/what this might break and how to mitigate.

I know that there are parts of the Filecoin ecosystem that assume there is always a single root and will break without it, so one tradeoff of this might be that those CAR files won’t work at all in Filecoin.

Reply:
>  there are parts of the Filecoin ecosystem

My understanding was that the FIL ecosystem only deals in "deterministic cars" and anything else is virtually unsupported. Whereas `.car` as a concept within `ipfs & friends` ecosystems is way less strict, and is essentially a "better `block put` interface". Is my approach to this wrong...?

Reply:
That’s not entirely accurate.

Filecoin clients that use the regular deal flow only use deterministic car files. The offline dealflow, or really any deal that is added to the network without our online dealflow code, supports any CAR file. The retrieval market **does not** require deterministic CAR files in order to read out data. It *does* require that there is a single root though.

I was also told that there are other actors in the Filecoin ecosystem reading CAR files that also assume there is always a single root and that multiple roots, in addition to no roots, would be problematic. In this case, I’m talking about developers outside of Protocol Labs building applications and tooling on top of the network.